cmd/nsm-sriov-manager/nsm-sriov-manager.go

```diff
@@ -0,0 +1,418 @@
+			Kind:       "ConfigMap",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "nsm-sriov-vf-list",
+			Namespace: "default",
+			Labels:    map[string]string{"networkservicemesh.io/nsm": "sriov-vf-list"},
+		},
+		Data: make(map[string]string),
+	}
+	dataString := make(map[string]string, 0)
+	for _, vf := range discoveredVFs.vfs {
+		data, err := yaml.Marshal(vf)
+		if err != nil {
+			return v1.ConfigMap{}, err
+		}
+		dataString[vf.pciAddr] = string(data)
+	}
+	configMap.Data = dataString
+
+	return configMap, nil
+}
+
+func checkVFIOModule() error {
+	// Checking if module vfio-pci is loaded
+	out, err := exec.Command("lsmod").Output()
+	if err != nil {
+		return fmt.Errorf("lsmod failed with error: %+v, generated output: %s", err, string(out))
+	}
+	if strings.Contains(string(out), "vfio_pci") {
+		logrus.Infof("vfio-pci kernel module has been found already loaded.")
+		return nil
+	}
+	// Attempting to load vfio-pci kernel module
+	out, err = exec.Command("modprobe", "vfio-pci").Output()
+	if err != nil {
+		return fmt.Errorf("modprobe failed with error: %+v, generated output: %s", err, string(out))
+	}
+	return nil
+}
+
+func buildVFIODevices(discoveredVFs *VFs) error {
+	if err := checkVFIOModule(); err != nil {
+		return err
+	}
+
+	for i, vf := range discoveredVFs.vfs {
+		logrus.Infof("VF: %s %+v", i, vf)
+		// Check if there is already vfio device for iommu group of VF
+		iommuGroup, err := strconv.Atoi(vf.iommuGroup)
+		if err != nil {
+			// Something wrong with iommu group for this VF
+			delete(discoveredVFs.vfs, vf.pciAddr)
+			logrus.Errorf("fail to convert iommu group with error: %+v for pci address: %s",
+				err, strings.Replace(vf.pciAddr, "-", ":", -1))
+			continue
+		}
+		vfioDevice := fmt.Sprintf("/dev/vfio/%d", iommuGroup)
+		_, err = os.Lstat(vfioDevice)
+		if !os.IsNotExist(err) && err != nil {
+			// Something wrong with access vfio path for iommu group
+			// it is safer to skip it and remove from the list of available VFs
+			// on the host.
+			delete(discoveredVFs.vfs, vf.pciAddr)
+			logrus.Errorf("fail to check for existing vfio device with error: %+v for pci address: %s",
+				err, strings.Replace(vf.pciAddr, "-", ":", -1))
+			continue
+		}
+		// vfio device for iommu group does not exist, need to create it
+		discoveredVFs.vfs[vf.pciAddr].VFIODevice = vfioDevice
+		if err := bindVF(vf); err != nil {
+			// Could not bind, cannot use, delete this VF
+			logrus.Errorf("fail to bind VF to vfio device with error: %+v for pci address: %s",
+				err, strings.Replace(vf.pciAddr, "-", ":", -1))
+			delete(discoveredVFs.vfs, vf.pciAddr)
+			continue
+		}
+	}
+	return nil
+}
+
+func waitAndRetry(timeout time.Duration, retries int, check func() bool) error {
+	ticker := time.NewTicker(timeout)
+	for r := 0; r < retries; r++ {
+		select {
+		case <-ticker.C:
+			if check() {
+				return nil
+			}
+		}
+
+	}
+
+	return fmt.Errorf("timeout has expired")
+}
+
+func bindVF(vf *VF) error {
```

Comment:
You really need to document what this tool is intended to do - code does not document intent (or we would never have bugs) and commit logs are not documentation.  It *could* be listing vfio-bound functions, VFs of PFs (any?  Some you list?)  It clearly *is* doing binding, but perhaps that should be an option?

Reply:
Sure thing, the doc is missing I will add it as soon as  we agree on what this tool should do and how ;)